sycl/doc/extensions/proposed/sycl_ext_oneapi_num_compute_units.asciidoc

```diff
@@ -0,0 +1,175 @@
+:lang: en
+:dpcpp: pass:[DPC++]
+:endnote: &#8212;{nbsp}end{nbsp}note
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2024 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+SYCL 2020 allows developers to query the maximum number of compute units in a
+device via the `info::device::max_compute_units` query.
+There are two issues with this existing query: first, that it refers to a
+"maximum", despite the number of compute units being a fixed property of a
+device; and second, that the definition of "compute units" is vague. Different
+implementations and backends do not provide consistent interpretations of this
+query, which makes it difficult for developers to use the number of compute
+units in a portable way.
+
+This extension provides a new query, `info::device::num_compute_units`, with
+the aim to clarify the meaning of "compute units" in SYCL and drive consistency
+across implementations.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_NUM_COMPUTE_UNITS` to one of the values defined in
+the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Compute units
+
+A SYCL device is divided into one or more compute units, which are each divided
+into one or more processing elements.
+
+All work-items in a given work-group must execute on exactly one compute unit,
+and the work-items within a work-group must not migrate between compute units
```

Comment:
I wonder if this part about "must not migrate" is too strong.  For example, the O/S could migrate O/S thread to different logical cores on the CPU implementation.  What is the rationale for including this prohibition in this spec?

Reply:
I was really just trying to come up with a stricter definition of "compute unit" to improve portability, by constraining implementation choice a little bit.  But it probably doesn't add anything: even the case you're worried about would still be permitted (because an implementation could choose to expose each O/S thread as a compute unit, and the property would still be satisfied as long as work-groups don't migrate between O/S threads).

Reply:
I took it out in https://github.com/intel/llvm/pull/16293/commits/100b35b47bc6e35f3cf0bcc3c185df854d4da9c9.